### PR TITLE
Facet clean up and a few more fixes 

### PR DESCRIPTION
**Pull request by kanitw**
Now we can get reasonable facet.

Note that we will to handle the following issues later:
- `paddingOuter` for point doesn't make (may need to just output `padding` if we don't wanna change vega https://github.com/vega/vega/issues/679
- The grid line is too long.
- Need to deal with axis grid layer
- If there is null value, the layout is still incorrect for both facet and non-facet plot. (This should be fixed in Vega https://github.com/vega/vega/issues/678)

vega_editor

cc: @domoritz

### CODE REVIEW

**Comment by kanitw** [comment]
Ah looks like I have to add more test on the plane lol

**Comment by kanitw** [comment]
@arvind  This should be ready for review.  Note that the only failing thing is the diff hit.  (only miss by 2%.) 

I think I added all the tests that are reasonable to test now.  Other stuffs are subject to change, so let's ignore the 2%.  :)

File: src/compile/facet.ts [hunk 1]
```diff
@@ -426,81 +426,65 @@ function parseAxisGroup(model: FacetModel, channel: Channel) {
 }
 
 
-function getXAxesGroup(model: FacetModel): VgMarkGroup {
-  const hasCol = !!model.facet().column;
+export function getSharedAxisGroup(model: FacetModel, channel: 'x' | 'y'): VgMarkGroup {
+  const isX = channel === 'x' ;
+  const facetChannel = isX ? 'column' : 'row';
+  const hasFacet = !!model.facet()[facetChannel];
```
Comment (domoritz) [inline src/compile/facet.ts]: `facetChannel in model.facet()`?

File: src/axis.ts [hunk 2]
```diff
@@ -98,6 +103,12 @@ export interface AxisConfig {
    * @minimum 0
    */
   subdivide?: number;
+
+  /**
+   * Whether the axis should include ticks.
```
Comment (domoritz) [inline src/axis.ts]: Include => use or show

File: src/compile/facet.ts [hunk 3]
```diff
@@ -21,6 +21,17 @@ import {assembleLayout, parseFacetLayout} from './layout';
 import {Model} from './model';
 import {initScale, parseScaleComponent} from './scale';
 
+/**
+ * Prefix for special data sources for driving column's axis group.
+ */
+
+export const COLUMN_AXES_DATA_PREFIX = 'column-';
```
Comment (domoritz) [inline src/compile/facet.ts]: We seem to be a bit inconsistent about whether we use - or _ for scale or data source names. I guess we can use - for data sources and _ for scales given that we use scales in expressions.

Comment (kanitw) [inline src/compile/facet.ts]: Good point.

File: src/compile/facet.ts (hunk 1 above)
Comment (kanitw) [inline src/compile/facet.ts]: But sometimes you want the whole object too, I don't want to make the return type inconsistent.

File: src/axis.ts (hunk 2 above)
Comment (kanitw) [inline src/axis.ts]: I think include seems fine here, especially given how many dependent PRs we have.

File: src/compile/facet.ts (hunk 1 above)
Comment (domoritz) [inline src/compile/facet.ts]: I don't get why you would have to change the type.

File: src/compile/facet.ts (hunk 3 above)
Comment (kanitw) [inline src/compile/facet.ts]: This one is a data source, so I guess I don't have to change anything.

File: src/compile/facet.ts (hunk 1 above)
Comment (kanitw) [inline src/compile/facet.ts]: `facet()` return the whole `Facet` object in FacetSpec.  

`facet(channel)` would return `FieldDef`.

**Pull request merged**